Modules/xASL_module_ASL.m

```diff
@@ -132,23 +144,147 @@
+    
+    if sum(x.Q.EchoTime<=0)>0
+        warning('x.Q.EchoTime should be positive');
     end
-    x.modules.asl.bMultiPLD = true;
+
+    % Handle different number of echo times (== potentially different ASL sequences)
+    if x.Q.NumberEchoTimes==1
+        fprintf('%s\n', 'Single-echo ASL detected');
+    elseif x.Q.NumberEchoTimes==2
+        warning('Dual-echo ASL detected, dual-echo ASL processing not yet implemented');
+        fprintf('%s\n', 'We will process these ASL images as if they have a single echo, using the first echo only');
+        fprintf('%s\n', 'Assuming that this is a dual-echo ASL+BOLD sequence');
+        x.Q.EchoTime = min(x.Q.EchoTime);
+    elseif x.Q.NumberEchoTimes>2
+        x.modules.asl.bMultiTE = true;
+        fprintf('%s\n', 'Multiple echo times detected, processing this as multi-echo ASL');
+        fprintf('%s\n', 'Note that this feature is still under development');
+    end
+    
+    fprintf('%s\n', 'Detected the following unique echo times (ms):')
+    for iTE = 1:x.Q.NumberEchoTimes-1
+        fprintf('%.2f, ', round(x.Q.uniqueEchoTimes(iTE),4));
+    end
+    fprintf('%.2f\n', round(x.Q.uniqueEchoTimes(end),4));
+end
+
+% Deal with too short vectors (e.g., repetitions in the case of single-PLD)
+nTE = length(x.Q.EchoTime);
+factorTE = nVolumes/nTE;
+x.Q.EchoTime = repmat(x.Q.EchoTime(:), [factorTE 1]);
+nTE = length(x.Q.EchoTime);
+
+% Number of echoes should now be equal to the number of ASL volumes
+if nTE~=nVolumes
+    warning(['Number of echo times (n=' xASL_num2str(nTE) ') should be equal to number of ASL volumes (n=' xASL_num2str(nVolumes) ')']);
+end
+
+
+%% D4. Multi-PLD parsing
+x.modules.asl.bMultiPLD = false; % by default, we use single-PLD processing
+
+if ~isfield(x.Q,'Initial_PLD')
+    warning('Missing field x.Q.Initial_PLD, this is needed for ASL quantification');
+elseif isempty(x.Q.Initial_PLD) || ~isnumeric(x.Q.Initial_PLD)
+    warning('Illegal field x.Q.Initial_PLD, this should not be empty and should contain numerical values');
 else
-    x.modules.asl.bMultiPLD = false;
+    x.Q.uniqueInitial_PLD = unique(x.Q.Initial_PLD);
+    x.Q.NumberPLDs = length(x.Q.uniqueInitial_PLD);
+    
+    if sum(x.Q.Initial_PLD<=0)>0
+        warning('x.Q.Initial_PLD should be positive');
+    end
+
+    % Handle different number of PLDs
+    if x.Q.NumberPLDs==1
+        fprintf('%s\n', 'Single-PLD ASL detected');
+    elseif x.Q.NumberPLDs>1
+        fprintf('%s\n', 'Multi PLDs detected, processing this as multi-PLD ASL');
+        fprintf('%s\n', 'Note that this feature is still under development');
+        x.modules.asl.bMultiPLD = true;
+    end
+
+    fprintf('%s\n', 'Detected the following unique PLDs (ms):')
+    for iPLD = 1:x.Q.NumberPLDs-1
+        fprintf('%d, ',x.Q.uniqueInitial_PLD(iPLD));
+    end
+    fprintf('%d\n',x.Q.uniqueInitial_PLD(end));
+end
+
+% Deal with too short vectors (e.g., repetitions in the case of single-PLD)
+nPLD = length(x.Q.Initial_PLD);
```

Comment:
The same error as above.

Reply:
Done in 0bdf54eb502121e43a25db8167411a91e183d051

Reply:
Same as above - we need error instead of a warning.

Reply:
Done in dc59f2a99a7cce88644bc78663b16cf426e823fa